[CARBONDATA-816]  Added Example for Hive Integration

anubhav100 opened the pull request:
Added Example For HIve Carbon Integration

anubhav100 commented:
@chenliang613 can you please have a look?

CarbonDataQA commented:
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1490/

CarbonDataQA commented:
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1495/

QiangCai commented on integration/hive/pom.xml:
```diff
@@ -95,7 +168,9 @@
                 <!-- Note config is repeated in scalatest config -->
                 <configuration>
                     <includes>
-                        <include>**/Test*.java</include>
```
add maven-scala-plugin, like spark2 module

anubhav100 replied:
changes done

QiangCai commented on integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala:
```diff
@@ -0,0 +1,146 @@
+      .appName("HiveExample")
+      .config("carbon.sql.warehouse.dir", warehouse).enableHiveSupport()
+      .getOrCreateCarbonSession(
+        "hdfs://localhost:54310/opt/carbonStore", metaStore_Db)
```
change to use local dir

anubhav100 replied:
@QiangCai changes done

QiangCai commented on integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala:
```diff
@@ -0,0 +1,146 @@
+    logger.info(s"============HIVE CLI IS STARTED ON PORT $port ==============")
+
+    stmt
+      .execute(s"ADD JAR $rootPath/assembly/target/scala-2.11/carbondata_2.11-1.1" +
```
check whether jar file is exists or not

anubhav100 replied:
@QiangCai  changes done

CarbonDataQA commented:
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1518/

anubhav100 commented:
@chenliang613 @QiangCai changes done can you have a look?

chenliang613 commented on integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala:
```diff
@@ -0,0 +1,166 @@
+        store, metaStore_Db)
+
+    val carbonJarPath = s"$rootPath/assembly/target/scala-2.11/carbondata_2.11-1.1" +
+                        s".0-incubating-SNAPSHOT-shade-hadoop2.7.2.jar"
```
For hadoop version of assembly jar , please don't give the fixed version number(2.7.2), how about using 2.* ?

anubhav100 replied:
@chenliang613  Wildcards are not supported in both hive.aux.jars.path and add jar.

i solved this problem by specifying both version of hadoop

val carbonHadoopJarPath = s"$rootPath/assembly/target/scala-2.11/carbondata_2.11-1.1" +
                              ".0-incubating-SNAPSHOT-shade-hadoop2.7.2.jar"

    val carbon_DefaultHadoopVersion_JarPath =
      s"$rootPath/assembly/target/scala-2.11/carbondata_2.11-1.1" +
      ".0-incubating-SNAPSHOT-shade-hadoop2.2.0.jar"

first i tried to add jar with 2.7.2 version if it fails then i tried with hadoop 2.2.0 version

 try {
      stmt
        .execute(s"ADD JAR $carbonHadoopJarPath")
    }
    catch {
      case exception: Exception =>
        logger.warn(s"Jar Not Found $carbonHadoopJarPath"+"Looking For hadoop 2.2.0 version jar")
        try {
          stmt
            .execute(s"ADD JAR $carbon_DefaultHadoopVersion_JarPath")
        }
        catch {
          case exception: Exception => logger
            .error(s"Exception Occurs:Neither One of Jar is Found $carbon_DefaultHadoopVersion_JarPath,$carbonHadoopJarPath"+"Atleast One Should Be Build")
            hiveEmbeddedServer2.stop()
        }
    }

CarbonDataQA commented:
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1529/

cenyuhai commented:
@anubhav100 can you create a new project under examples and move these codes?
examples
------flink
------spark
------spark2
------hive
---------your codes

cenyuhai commented on integration/hive/pom.xml:
```diff
@@ -64,6 +64,79 @@
+            <version>4.3-alpha1</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.hadoop</groupId>
```
hadoop-client already contains hadoop-hdfs and hadoop-common, what hadoop version do you want?

anubhav100 replied:
done

chenliang613 commented:
@cenyuhai  HiveExample can more to examples/ later once hive integration module be completely moved into master.

cenyuhai commented on integration/hive/pom.xml:
```diff
@@ -64,6 +64,79 @@
         <dependency>
+            <groupId>org.apache.spark</groupId>
+            <artifactId>spark-network-common_2.10</artifactId>
+            <version>2.1.0</version>
```
carbondata-spark2 has already included this dependency

anubhav100 replied:
resolved

cenyuhai commented on integration/hive/pom.xml:
```diff
@@ -64,6 +64,79 @@
+        <dependency>
+            <groupId>org.apache.hadoop</groupId>
+            <artifactId>hadoop-hdfs</artifactId>
+            <version>2.7.3</version>
```
it is already included in carbondata-hadoop

anubhav100 replied:
resolved

cenyuhai commented on integration/hive/pom.xml:
```diff
@@ -64,6 +64,79 @@
+        <dependency>
+            <groupId>org.apache.spark</groupId>
+            <artifactId>spark-repl_${scala.binary.version}</artifactId>
+        </dependency>
```
it is already included in carbondata-spark2

anubhav100 replied:
@cenyuhai resolved comments

cenyuhai commented:
I test this example in my computer, it's very good. But I think you should remove some unnecessary dependencies.

chenliang613 commented on integration/hive/src/main/scala/org/apache/carbondata/hiveexample/HiveExample.scala:
```diff
@@ -0,0 +1,179 @@
+
+    carbon
+      .sql(
+        """create table hive_carbon_example (id int,name string,salary double) stored by
```
Please use uppercase for DDL and DML key words.

anubhav100 replied:
@chenliang613 resolved

CarbonDataQA commented:
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1553/

anubhav100 commented:
@cenyuhai @chenliang613 i resolved all the comments can please have a look?can it be merged?

CarbonDataQA commented:
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1557/

chenliang613 commented:
@anubhav100  everything looks good, can you squash all commits into one commit.

When ending the HiveExample, the application can't be automatically stopped, we must to manually stop it:
<img width="711" alt="2017-04-12 01 20 06" src="https://cloud.githubusercontent.com/assets/8075709/24927980/69846f54-1f1e-11e7-8e3a-b9a64dbec034.png">

anubhav100 commented:
@chenliang613 i added the code for stoping application automatically ,tested it on my system and rebased all commits into one commit can it be merged?

CarbonDataQA commented:
Build Success with Spark 1.6.2, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder/1579/

chenliang613 commented:
@anubhav100  Thanks for your good contribution.

LGTM
